modules/logging.py

```diff
@@ -0,0 +1,36 @@
+# app core modules
+from typing import TYPE_CHECKING, TypedDict
+
+from modules.app.read_write import ReadWrite
+from modules.app.settings import Settings
+from modules.tcp import TCP
+
+import datetime;
+
+if TYPE_CHECKING:
+    from main import Application
+
+class Logging:
+    class LogData_t(TypedDict):
+        ip: str
+        file: str
+        comment: str
+        datetime: str
+
+    def __init__( self, context ) -> None:
+        self.context : "Application" = context
+        self.settings : Settings = context.settings
+
+    def log( self, data : LogData_t ):
+        """Write to log file"""
+        log = self.context.read_write.getLogFile()
+        log.append( data )
+        self.context.read_write.writeLogFile( log )
+
+    def log_file( self, filename : str, comment : str ):
+        data : self.LogData_t = { "ip": self.settings.server_ip, 
```

Comment:
This should be changed to the following (not entirely sure why it works that way, though):
```suggestion
        data : Logging.LogData_t = { "ip": self.settings.server_ip, 
```

Reply:
self. works because ```LogData_t``` is a nested type, to reference it from other classes ```Logging.LogData_t``` is required.
using self keeps it concise. Both are valid though.

Reply:
I suggested the change, as VS Code was throwing an error. I was a bit surprised about that too, as one would think using `self` should work. Here is what ChatGPT has to say about it:

> The error occurs because `self.LogData_t` is interpreted as an attempt to access an instance attribute, not a nested class. In Python, nested classes are accessed using the class name (e.g., `Logging.LogData_t`), not through an instance (`self`).
> 
> Here’s why:
> 
> 1. **Scope of Nested Classes**: The `LogData_t` class is defined inside the `Logging` class. This makes it a **class-level attribute** of `Logging`. It cannot be accessed through an instance (`self`) unless explicitly set as an instance attribute (which it is not in your code).
> 2. **`self` Refers to the Instance**: `self` refers to the current instance of the `Logging` class. It does not automatically have access to nested classes or class-level attributes unless explicitly assigned.